File: code/ch17/17.2.3.largestRectangleArea.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        l, r, ans = [-1] * n, [n] * n, 0

        for i in range(1, n):
            j = i - 1
            while j >= 0 and heights[j] >= heights[i]:
                j = l[j]
            l[i] = j
        for i in range(n - 2, -1, -1):
            j = i + 1
            while j < n and heights[j] >= heights[i]:
                j = r[j]
            r[i] = j
        for i in range(n):
            ans = max(ans, heights[i] * (r[i] - l[i] - 1))
        return ans

File: code/ch17/17.2.3.largestRectangleArea.java
import java.util.*;
import java.lang.Math;

public class Solution {
  public int largetstRectangleArea(int[] heights) {
    int n = heights.length, ans = 0;
    List<Integer> l = new ArrayList<>(),
      r = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      l.add(-1);
      r.add(n);
    }
    for (int i = 0; i < n; i++) {
      int j = i - 1;
      while (j >= 0 && heights[j] >= heights[i])
        j = l.get(j);
      l.set(i, j);
    }
    for (int i = n - 2; i >= 0; i--) {
      int j = i + 1;
      while (j < n && heights[j] >= heights[i])
        j = r.get(j);
      r.set(i, j);
    }
    for (int i = 0; i < n; i++)
      ans = Math.max(ans, heights[i] * (r.get(i) - l.get(i) - 1));
    return ans;
  }
}
